src/category_theory/sites/sheaf.lean

```diff
@@ -54,13 +54,97 @@ variables (J : grothendieck_topology C)
-presheaf of types given by sending U : C to Hom_{A}(X, P U) is a sheaf of types.
+A sheaf of A is a presheaf P : Cᵒᵖ => A such that for every E : A, the
+presheaf of types given by sending U : C to Hom_{A}(E, P U) is a sheaf of types.
 
 https://stacks.math.columbia.edu/tag/00VR
 -/
 def is_sheaf (P : Cᵒᵖ ⥤ A) : Prop :=
-∀ X : A, presieve.is_sheaf J (P ⋙ coyoneda.obj (op X))
+∀ E : A, presieve.is_sheaf J (P ⋙ coyoneda.obj (op E))
+
+section limit_sheaf_condition
+open presieve presieve.family_of_elements limits
+variables (P : Cᵒᵖ ⥤ A) {X : C} (S : sieve X) (R : presieve X) (E : Aᵒᵖ)
+
+@[simps] def cones_equiv_sieve_compatible_family :
+  (functor.cones (S.arrows.diagram.op ⋙ P)).obj E ≃
+  {x : family_of_elements (P ⋙ coyoneda.obj E) S // x.sieve_compatible} :=
+{ to_fun := λ π, ⟨λ Y f h, π.app (op ⟨over.mk f, h⟩), λ _, by
+    { intros, apply (id_comp _).symm.trans, dsimp,
+      convert π.naturality (quiver.hom.op _),
+      refl, refl, refl, swap, apply over.hom_mk, refl }⟩,
+  inv_fun := λ x, { app := λ f, x.1 f.unop.1.hom f.unop.2,
+    naturality' := λ f f' g, by
+    { refine eq.trans _ (x.2 f.unop.1.hom g.unop.left f.unop.2),
+      erw id_comp, congr, rw over.w g.unop } },
+  left_inv := λ π, by { ext, dsimp, congr,
+    rw op_eq_iff_eq_unop, ext, symmetry, apply costructured_arrow.eq_mk },
+  right_inv := λ x, by { ext, refl } }
+
+variables {P S E} {x : family_of_elements (P ⋙ coyoneda.obj E) S} (hx : x.sieve_compatible)
+
+@[simp] def _root_.category_theory.presieve.family_of_elements.sieve_compatible.cone :
+  cone (S.arrows.diagram.op ⋙ P) :=
+{ X := E.unop, π := (cones_equiv_sieve_compatible_family P S E).inv_fun ⟨x,hx⟩ }
+
+def hom_equiv_amalgamation :
+  cone_morphism hx.cone (P.map_cone S.arrows.cocone.op) ≃ {t // x.is_amalgamation t} :=
+{ to_fun := λ l, ⟨l.hom, λ Y f hf, l.w (op ⟨over.mk f, hf⟩)⟩,
+  inv_fun := λ t, ⟨t.1, λ f, t.2 f.unop.1.hom f.unop.2⟩,
+  left_inv := λ l, by { ext, refl },
+  right_inv := λ t, by { ext, refl } }
+
+variables (P S)
+lemma is_limit_iff_is_sheaf_for :
+  nonempty (is_limit (P.map_cone S.arrows.cocone.op)) ↔
+  ∀ E : Aᵒᵖ, is_sheaf_for (P ⋙ coyoneda.obj E) S :=
+begin
+  dsimp [is_sheaf_for], simp_rw compatible_iff_sieve_compatible,
+  rw ((cone.is_limit_equiv_is_terminal _).trans (is_terminal_equiv_unique _ _)).nonempty_congr,
+  rw classical.nonempty_pi, split,
+  { intros hu E x hx, specialize hu hx.cone,
+    erw (hom_equiv_amalgamation hx).unique_congr.nonempty_congr at hu,
+    exact (unique_subtype_iff_exists_unique _).1 hu },
+  { rintros h ⟨E,π⟩, let eqv := cones_equiv_sieve_compatible_family P S (op E),
+    rw ← eqv.left_inv π, erw (hom_equiv_amalgamation (eqv π).2).unique_congr.nonempty_congr,
+    rw unique_subtype_iff_exists_unique, exact h _ _ (eqv π).2 },
+end
+
+lemma subsingleton_iff_is_separated_for :
+  (∀ c, subsingleton (c ⟶ P.map_cone S.arrows.cocone.op)) ↔
+  ∀ E : Aᵒᵖ, is_separated_for (P ⋙ coyoneda.obj E) S :=
```

Comment:
Similarly here, I think we have a named proposition called `is_separated` which is essentially this.

Reply:
Again, if we quantify over all S then we can use `is_separated` for `SheafOfTypes` in the second proposition, though for general sheaves, `is_separated` isn't defined, so I can only state the equivalence as:
```
lemma subsingleton_iff_is_separated :
  (∀ E : Aᵒᵖ, is_separated J (P ⋙ coyoneda.obj E)) ↔
  ∀ ⦃X : C⦄ (S : sieve X), S ∈ J X → ∀ c, subsingleton (c ⟶ P.map_cone S.arrows.cocone.op) :=
```
Maybe I should add the def `is_separated` for general sheaves? I think it's used in your work on sheafification as "hsep", but hsep seems to be for concrete categories only.